Refactor dependency handling for the Braze SDK

tjmw opened the pull request:
## What does this change?

Refactors the way dependencies for loading the Braze web SDK are handled. By dependencies we mean the information we need and the conditions which need to be true before we can load the SDK. Currently these are:

* That the switch is on
* The API key is set
* The relevant consents have been given
* That the page isn't paid (GLabs) content
* That the user is signed in and we have their uuid
* That the user _is_ a Gu supporter

### Before

Dependency checks were all intermingled with other code in `BrazeBanner`.

### After

Dependency checks have been extracted out to `checkBrazeDependencies.ts` and are configured in a way that we think makes them much clearer. We're using the new (generic) `dependecyChecker` which we've added in this PR, which takes an array of async dependencies/conditions and processes them until one fails. Note that this is a slight change from the original code where we would make each check in turn in serial, now we're kicking them all off "concurrently" (in quotes because JS is single threaded. Some of the work _may_ be parallelisable, for example the request to idapi) and processing the results serially, until we come across a check which fails.

## Why?

* It's hopefully clearer to a new/returning reader of the code what the dependencies actually are.
* It's easier to write tests around the dependencies.
* We'd like to extract more Braze related code out to its own NPM package, which could be shared across DCR and frontend. Modularising the existing code is a first step along that path.
* It's much easier to have functionality like common logging for when dependencies aren't all met which is really useful when developing.

tjmw commented on src/web/lib/braze/checkBrazeDependencies.test.ts:
```diff
@@ -0,0 +1,242 @@
+	return flushPromises;
+});
+
+let mockBrazeUuid: string | null;
```
I don't ❤️  all the mocking we're doing here but I think it's worth it for the confidence that the wiring etc is all correct, and at least it's all confined to this one test. This is much more of an integration test than a unit test.

liywjl commented on src/web/lib/dependencyChecker.ts:
```diff
@@ -0,0 +1,68 @@
+				};
+			}
+
+			return {
+				isSuccessful: true,
+				data: {
+					[cur.name]: result,
+					...syncAcc.data,
+				},
+			};
```
feels like it would be clearer if this retrun was in the `try` block

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,63 @@
+	isSignedIn: boolean,
+	idApiUrl: string,
+): Promise<Array<DependencyConfig>> => {
+	return [
+		{
```
I suspect prettier is not being applied correctly here.

tjmw replied:
Yeah I wondered the same thing, and had a conversation with Alex S on Friday about it. I _think_ it's just GitHub rendering the diff weirdly. Locally it looks fine and I see the correct number of tab characters. I'll check again to make sure.

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,63 @@
+import { hasRequiredConsents } from './hasRequiredConsents';
+import { hideSupportMessaging } from './hideSupportMessaging';
+
+const getCompleteDependencies = async (
```
the function is using `async`, but you are not using `await` anywhere

tjmw replied:
Good spot, this is fixed in the latest commit I pushed.

liywjl reviewed (commented):
I fully agree in the direction you are talking the Braze implementaion, I think it will help simplilfy and make it more testable in the future. 

However I think we could look at simplifying some of the logic around `checkBrazeDependencies.ts` and `dependencyChecker.ts`. I feel like a lot of what you are trying to achieve could be handled by `Promise.all`. Doing so would be a lot easier to understand to new devs looking at the code as they would likely understand the mental model quicker that using a `reduce` to handle each case. 

I am unsure what is the best approch would be, but if Braze requires all conditions to be met before displaying, maybe encapsulating each condtion in a promise wrapped by `Promise.all` might help.

tjmw commented:
Hi @liywjl thank you for the feedback! We've also had a discussion within the team about whether the `dependencyChecker` code is worthwhile given the complexity it adds, so I think your concerns are totally valid. I've just pushed up a commit which uses a simpler approach, removing `dependencyChecker` completely. I'd be interested in your thoughts on the change (we'll also discuss within TX).

tjmw commented on src/web/lib/braze/checkBrazeDependencies.test.ts:
```diff
@@ -0,0 +1,242 @@
+	windowSpy = jest.spyOn(window, 'window', 'get');
+});
+
+afterEach(() => {
+	windowSpy.mockRestore();
+});
```
Good spot, I've got that on a mental todo list of things to fix! Makes sense to refactor now.

tjmw commented:
Hi @liywjl I think I've addressed the feedback you left, and have moved away from using `dependencyChecker.ts` to a simpler approach. I'd be interested to hear what you think about the latest changes, particularly the stuff in 105277b.

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -1,63 +1,96 @@
-	);
+	const data: ResultData = {};
+
+	{
+		const name = 'brazeSwitch';
+		const value = window.guardian.config.switches.brazeSwitch;
+		if (!value) {
+			return buildFailureResponse(name, value, data);
+		}
+		data[name] = value;
+	}
```
The structure I has in mind was more on the lines of:

```
const brazeSwitch = new Promise((resolve, reject) => {
    const value = window.guardian.config.switches.brazeSwitch;
    if (!value) {
        reject(buildFailureResponse(name, value, data))
    }
    resolve(value)
});

[...]

return [brazeSwitch, ...]
```

That means when you call `checkBrazeDependencies` you can do something like:

```
const got = await Promise.all[checkBrazeDependencies(isSignedIn, idApiUrl)}
    .then(results => {
        ...
    })
    .catch(err => {
        ...
    });
```

Also in thise case it might be worth renaming `checkBrazeDependencies` to something like `arrayBrazeChecks` (not fully convinced by the my suggestion either)

The example I based on a post I found really useful https://www.javascripttutorial.net/es6/javascript-promise-all/
Thoughts on this?

tjmw replied:
Thanks for the detail! I considered `Promise.all` before, but ended up discounting it as the behaviour is _slightly_ different to what we have now. Currently the failure (if any) will be for the first item in the ordered list of dependencies. Using `Promise.all` changes that slightly to be the first item which fails chronologically, which _might_ not be the same. Especially if some operations are longer running than others (for example an API call). I'm now wondering how relevant or important that is though.

The other piece I'm not sure about with `Promise.all` is that in the event that one of the checks fails, we want to be able to access the data from any other successful checks. I don't think `Promise.all` lets you access that data (which is why we were reducing a list of promises in the `dependencyChecker`).

I definitely agree that it would be nice to be using common JavaScript APIs and building blocks here. I'll try some things out and report back.

liywjl replied:
I see, maybe it might be worth looking at using `Promise.allSettled()` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

From what I gather that might be a better than `Promise.all ` for what you are hoping to achieve?

tjmw replied:
Hi @liywjl I had a play with `Promise.allSettled` but wasn't super happy with the result (it turns out it isn't supported in the version of Node we're on which adds some additional friction). I pushed up a few more commits, including f160afb, which is kind of a compromise approach where I iterate over the conditions rather than pass them to `Promise.allSetted`. I think it could be a nice happy medium but interested in your take on it.

I left the `allSetted` approach in there but commented out so you can see how it looked. The part where we iterate the results to build up the final `DependenciesResult` object wasn't very nice.

tjmw replied:
Hi @liywjl I've had a bit more of a play around with refactoring this but didn't get to a point where I was happier with the result than what's there now. I've removed the commented `Promise.allSettled` experimental code and have left a comment on using `Promise.allSettled` in future when it's more reliably available. What are your thoughts on me merging this as it is?

liywjl reviewed (commented):
Awesome work! This is definitely an improvement, still had another suggestion on how `checkBrazeDependencies`, curious to know your thoughts on my suggestion?

liywjl commented on src/web/components/StickyBottomBanner/BrazeBanner.tsx:
```diff
@@ -252,35 +220,35 @@ export const canShow = async (
+		const { failureData, failureField, data } = dependenciesResult;
+		if (SDK_OPTIONS.enableLogging) {
+			// eslint-disable-next-line no-console
+			console.log(
```
`console.error` might be better?

tjmw replied:
It's actually not super exceptional to end up here, so an error doesn't feel quite right to me (for example, any user who isn't logged in won't meet all the dependencies, so will end up here). This is mostly useful to a developer working on the code (also in the wild `SDK_OPTIONS.enableLogging` won't be enabled).

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,100 @@
+
+			if (result) {
+				data[name] = result;
+			} else {
+				return buildFailureResponse(name, result, data);
+			}
```
maybe just throw an error here, so `buildFailureResponse` only ever used in catch block?

liywjl replied:
Also if `buildFailureResponse` is only defined once, then maybe we don't need it as a seperate function? we could return the object here?

tjmw replied:
Ah interesting question! This did actually cross my mind, but I decided against (using exceptions for flow control never feels quite right to me). Maybe the scope here is tight enough that it's not a problem though. It would be nice to have a single failure case build in one place only. I'll give it a try and see how it looks!

tjmw replied:
I did try this but it felt clearer to me keeping the falsy case and the catch separate.

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,100 @@
+		},
+		{
+			name: 'brazeUuid',
+			value: isSignedIn ? getBrazeUuid(idApiUrl) : Promise.resolve(null),
```
worth wrapping in a Promise?

oliverlloyd replied:
I might be wrong but I think you can just use the `async` key to do this?

tjmw replied:
So I guess wrapping in a promise would look like this, is this what you had in mind?

```js
new Promise((resolve) => {
  if (isSignedIn) {
    getBrazeUuid(idApiUrl).then(resolve);
  } else {
    resolve(null);
  }
})
```

> I might be wrong but I think you can just use the async key to do this?

I'm interested in this! I _thought_ you could only use `async` to mark a function as async (which as you say means the function will always return a promise). Do you mean use an immediately invoked function in combination with `async` here?

oliverlloyd replied:
It was a drive by comment without fully parsing the code in my head but yes, `async` is like shorthand for 'this function always returns a promise' which I though might be handy here but take my idea with a lot of salt!

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,100 @@
+	name: string;
+	value: Promise<string | boolean | null | undefined>;
+};
+const buildFailureResponse = (name: string, value: any, data: ResultData) => ({
+	isSuccessful: false,
+	failureField: name,
+	failureData: value,
+	data,
+});
```
personal preferrence but maybe the response could be structured like:

```
{
    isSuccessful: false,
    error: {
         failureField: name,
	failureData: value,
    },
    data
}
```

liywjl replied:
that way if there are no errors we could have something that looks like:

```
{
    isSuccessful: true,
    error: null,
    data
}
```

Its nicer to not have to check if an object key exists and always assume it does exist  but is not set

tjmw replied:
Oh interesting, yeah combining the two failure fields together is nice. Maybe the types as they are right now are overdoing it a bit. Currently I've got:

```js
type SuccessResult = {
	isSuccessful: true;
	data: ResultData;
};

type FailureResult = {
	isSuccessful: false;
	failureField: string;
	failureData: any;
	data: ResultData;
};

type DependenciesResult = SuccessResult | FailureResult;
```

But given a single nullable error key this could become:

```js
type DependencyError = {
        failureField: string;
	failureData: any;
};

type DependenciesResult = {
        error: DependencyError | null;
	data: ResultData;
}
```

The consumer would then just check if `error` is truthy to know the overall outcome.

> Its nicer to not have to check if an object key exists and always assume it does exist but is not set

I think in this case TypeScript is helping us, since if `isSuccessful` is `false` the type of `FailureResult` specifies that the `failureField` and `failureData` fields will always be defined. (Interesting blog post describing the approach I used here: https://mariusschulz.com/blog/more-literal-types-in-typescript#boolean-literal-types)

tjmw replied:
I went ahead and combined the two failure fields into one object which I agree is nicer, thanks for the suggestion! For now I've kept separate `SuccessResult` and `FailureResult` types.

liywjl commented on src/web/lib/braze/checkBrazeDependencies.ts:
```diff
@@ -0,0 +1,100 @@
+		}
+	}
+
+	return {
+		isSuccessful: true,
+		data,
+	};
```
ditto about returning something in the structure like:

```
{
    isSuccessful: true,
    error: null,
    data
}
```

liywjl reviewed (approved):
LGTM 👍 all the comments are just personal preferences, so feel free to ignore if you disagree
